### ISSUE TITLE
Dependency on System.Runtime.InteropServices causes runtime error for apps targeting net451 on *nix systems (i.e. running on Mono)

### ISSUE THREAD

**Issue by JunTaoLuo** [dotnet/corefx#9012]
## Scenario

1) Running a `net451` app depending on `System.Runtime.InteropServices` on OSX/Ubuntu will produce the following exception at runtime:
```
 dotnet -v run -f net451                                                                                                                                                                               1 ↵
Telemetry is: Enabled
Project RuntimeServiceOnMono (.NETFramework,Version=v4.5.1) was previously compiled. Skipping compilation.
Running /usr/local/bin/mono --debug /Users/jtluo/Documents/workspace/juntaoluo/tp/RuntimeServiceOnMono/bin/Debug/net451/osx.10.11-x64/RuntimeServiceOnMono.exe
Process ID: 29130

Unhandled Exception:
System.DllNotFoundException: System.Native
  at (wrapper managed-to-native) Interop/Sys:GetUnixNamePrivate ()
  at Interop+Sys.GetUnixName () [0x00000] in <filename unknown>:0
  at System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform (OSPlatform osPlatform) [0x00000] in <filename unknown>:0
  at ConsoleApplication.Program.Main (System.String[] args) [0x00000] in <filename unknown>:0
[ERROR] FATAL UNHANDLED EXCEPTION: System.DllNotFoundException: System.Native
  at (wrapper managed-to-native) Interop/Sys:GetUnixNamePrivate ()
  at Interop+Sys.GetUnixName () [0x00000] in <filename unknown>:0
  at System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform (OSPlatform osPlatform) [0x00000] in <filename unknown>:0
  at ConsoleApplication.Program.Main (System.String[] args) [0x00000] in <filename unknown>:0
```

## Example
See repro at https://github.com/JunTaoLuo/RuntimeServiceOnMono. To run the sample, run `dotnet restore` and `dotnet run -f net451`

I understand Mono is not a scenario that's actively being developed for but we should not be causing a exception like this especially since we intend to use the `InteropServices` APIs in often used ASP.NET packages like Logging. 

Currently this means that most of our samples and apps cannot run on Mono (full CLR on *nix).

## Environment
`dotnet --version`:
```
Microsoft .NET Core Shared Framework Host

  Version  : 1.0.1-rc3-004312-00
  Build    : 1db6c07638a70a621b312e78d4dc9fb31a530f2f
```
`mono -V`:
```
Mono JIT compiler version 4.0.5 ((detached/1d8d582 Tue Oct 20 15:15:33 EDT 2015)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  x86
	Disabled:      none
	Misc:          softdebug
	LLVM:          yes(3.6.0svn-mono-(detached/a173357)
	GC:            sgen
```

cc @BrennanConroy @muratg @Eilon

**Comment by joshfree** [comment]
@JunTaoLuo could you upgrade to CoreFX rc3-24128-00 / Microsoft.NetCore.App 1.0.0-rc3-004338 / CLI 1.0.0-preview2-002911 and report back?

**Comment by mellinoe** [comment]
I suspect this will still be happening on the latest packages. BTW, it looks like the issue is with System.Runtime.InteropServices.RuntimeInformation. We have a `net45` implementation for Windows, but not for Unix. Unix probably always picks up the `netstandard1.1` version, which won't necessarily work on Mono. I guess it would work if System.Native was deployed with the app, but that probably won't happen by default.

**Comment by muratg** [comment]
cc @Eilon @DamianEdwards @davidfowl

**Comment by ericstj** [comment]
Updated the title and assignee.  This isn't for Interop, it's RuntimeInformation.  I would expect some different results using the latest packages, since those will probably result in the desktop implementation being chosen.

Not much we can do in packaging here since Mono pretends to be desktop and doesn't use a RID.

We could make the desktop library try to p-inoke to some *nix APIs (directly, without shim) if it didn't find the Windows ones.  Not sure how folks feel about that.

**Comment by JunTaoLuo** [comment]
I still see the same error with the latest packages. `System.Runtime.InteropServices.RuntimeInformation/4.0.0-rc3-24128-00`. CLI version: `1.0.0-preview2-002913`. Note that the repro only targets `net451` and does not depend on `Microsoft.NetCore.App`.

**Comment by joshfree** [comment]
@ellismg is going to take a look

**Comment by ellismg** [comment]
https://github.com/dotnet/cli/issues/3350 is a CLI problem that is causing a unix asset to be deployed instead of the Windows one.  Even if we deploy the correct asset, we still end up with p/invokes into kernel32 (which won't work) but we'd at least be able to provide some light-up mono implementation via late-binding.

**Comment by sunsided** [comment]
btw., same on mono 4.5.2.372 on Ubuntu 14.04, targeting `net452` with `1.0.0` packages from NuGet. I see it on the console logger though (`.AddConsole()`).

Is there some workaround we can use?

**Comment by borgdylan** [comment]
I confirm that by using a separately built System.Native, I could get the RuntimeInformation class to work on Mono on both x86 and x64 Ubuntu 16.04. It is a matter of shipping those libraries as source/binaries for the architectures that mono supports.

**Comment by borgdylan** [comment]
@ericstj The .NET CLI includes a RuntimeInformation class that detects Windows vs. *nix cases and P/Invokes straight thru to the native syscalls. I confirm that, that code cross-compiles and works fine unmodified on Linux/Mono: https://github.com/dotnet/cli/tree/rel/1.0.0/src/Microsoft.DotNet.InternalAbstractions/Native

**Comment by ericstj** [comment]
@borgdylan that's already understood.  As @ellismg mentioned we need to decide how we want to do that since mono represents itself as .NET desktop WRT to NuGet.

**Comment by borgdylan** [comment]
The way the CLI internal abstractions work would be best. That code works ok irrespective of CLR and treats operating systems directly. The alternative would be to make mono use parts of System.Native and friends as part of its codebase so that APIs new in corefx and not present in mono would just work. That would open up prospects of using mono's core parts i.e its CLR as a CoreCLR stand-in on platforms were the CoreCLR does not run in order to run apps that target corefx. Example platforms are Linux x86, LInux on MIPS, Linux on PowerPC.

**Comment by borgdylan** [comment]
The latter case would need cooperation from the Xamarin team.

**Comment by ericstj** [comment]
> code works ok irrespective of CLR

Yep that's what we're thinking to do for this particular case. We'll make the desktop implementation of this contract "work" on Mono and give some meaningful results without relying on p-invokes to a native shim we cannot deploy.  I suspect we'll do some reflection lightup to detect it's mono then call some mono APIs that provide similar information.  That's what @ellismg meant by late binding.

I don't think we want to have the native shims become something that's part of Mono.  They are too brittle.

**Comment by borgdylan** [comment]
@ericstj The reflection based search for ```Mono.Runtime``` where finding it means Mono and not finding it means .NET CLR has been done for years without issues. I consider this as the Litmus test for Mono.

**Comment by ericstj** [comment]
Yep, the other thing we need is an implementation of [this](https://github.com/dotnet/corefx/blob/master/src/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs) for Mono, ideally without any PInvokes.  Perhaps we can translate mono's `Environment.OSVersion` if that's reliable.

**Comment by borgdylan** [comment]
First of all due to how full framework behaves, Mono will not distinguish between Linux and Mac in the Environment APIs. Both will be reported as ```Unix```.

**Comment by borgdylan** [comment]
The OS version will be the Linux kernel version.

**Comment by borgdylan** [comment]
Therefore this cannot be done using pure Mono/.NET without P/Invokes. There is a reason why System.Native was used for CoreCLR. The only way to make this API run perfect is to have native implementation libraries on all three CLRs or to have all three CLRs ship a runtime info file or use an RID.

**Comment by borgdylan** [comment]
The code in: https://github.com/dotnet/cli/blob/rel/1.0.0/src/Microsoft.DotNet.InternalAbstractions/Native/PlatformApis.cs solves part of the problem but requires IO and P/Invokes.

**Comment by borgdylan** [comment]
The code in: https://github.com/dotnet/cli/blob/rel/1.0.0/src/Microsoft.DotNet.InternalAbstractions/RuntimeEnvironment.cs reiterates the fact that you cannot distinguish Intel/ARM/MIPS/PowerPc. Mono supports a host of architectures and you cannot get it right without the System.Native port to all those architectures.

**Comment by borgdylan** [comment]
@ericstj Do remember that Mono means Xamarin which means MSFT. I do see a world where .NET and mono teams collaborate that have a subset of System.Native be present in Mono and be compatible with the CoreCLR version.

**Comment by ericstj** [comment]
> Do remember that Mono means Xamarin which means MSFT.

@borgdylan I'm not arguing here, nor do I disagree with anything that's been said around mono support.  I am sorry if that's how this is coming off, but that's the last thing on my mind.  I want to support mono and I want to do so with minimal coupling.  I'm just stating the constraints we have and discussing possible design choices.

>you cannot get it right without the System.Native port to all those architectures.

Shims don't solve the processor architecture problem, they just push it to NuGet.  The shims [hardcode](https://github.com/dotnet/corefx/blob/d0dc5fc099946adc1035b34a8b1f6042eddb0c75/src/Native/System.Native/pal_runtimeinformation.cpp#L37) the architecture value and NuGet uses RID to give us the shim with the right hardcoded value.  I'd like to keep the goal of not having mono projects require a RID.  

Which I understand is why you are suggesting:
>System.Native be present in Mono

/cc @stephentoub @jkotas @davidfowl 
The shims are not public surface area.  They version quite frequently and are required to do so in order to support new API.  If we put them in mono it means mono *pins* itself to a specific shim version that must be target-able by upstack packages.  We don't have any way to represent that today.  It means we'd have to 'know' that net46 means a mono version with shim version X and if a new implementation required version X+1 of that shim it couldn't be supported on net46 due to mono's pinning, even though that has nothing to do with desktop framework.  I really don't want to go down this rabbit hole if it can be avoided.

The main reason we have shims was to avoid many forks of the IL per-distro/version, initially we were doing all the translation in IL and directly PInvoking to the *nix OS.  I suspect there is a way to do so here, we just need someone more knowledgeable in *nix than myself to do a deep dive.  I think @ellismg already has an idea here so I'd like to wait for his feedback, especially since this issue is assigned to him, before we go further down the path of trying to design this.

**Comment by borgdylan** [comment]
Ok. I just cannot understand how you would get the actual architecture without Mono and .NET having a way to tell that in the first place. In fact the problem arises from the fact that any CLR is there to abstract the CPU architecture from the IL code which is architecture independent. This is what the S.R.IS.RI API is for and therefore this is a chicken and egg type problem.

**Comment by ericstj** [comment]
We could do something like we do on [Windows](https://github.com/dotnet/corefx/blob/2465a77fb43b5380607f22f62671ddd3aa7dbbdb/src/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Windows.cs#L50) where we make a platform call to do it.  Just need to find the right method that isn't too distro specific.  Perhaps reading `/proc/cpuinfo` is enough.

**Comment by borgdylan** [comment]
That only lists each CPU core and its capabilities. It makes no mention of the architecture. Calling ```uname -a``` will yield a better result.

```
dylan@ubuntu-server:~$ uname -a
Linux ubuntu-server 4.4.0-28-generic #47-Ubuntu SMP Fri Jun 24 10:09:13 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```
On a Mac, LInux becomes Darwin and the architecture constant may be different. For Linux x86_64 means x64, i386 and i686 mean x86 and armv7l means ARM. I never saw the architecture constant for Aarch64.

**Comment by borgdylan** [comment]
Some docs: http://linux.die.net/man/1/uname

**Comment by borgdylan** [comment]
@JunTaoLuo The issue does not only appear when using Logging but also when using Kestrel in general if you target RTM.

**Comment by Tragetaschen** [comment]
Does uname handle / show the difference between hard-float and soft-float on ARM?

**Comment by borgdylan** [comment]
Do you know whether you have an armel or amhf install?

**Comment by borgdylan** [comment]
It should. amrv7l means hard float. Soft float may be shown as armv5.

**Comment by borgdylan** [comment]
This: https://wiki.debian.org/Multiarch/Tuples says that I am wrong on my latest comment. They will both show up as ```arm```. Since the hard/soft float issue is serious we need another command that can distinguish between them.

**Comment by Tragetaschen** [comment]
Usually you have a compiler tool chain that has built your Linux environment and that emits either hard float or soft float code. So in a .NET world: couldn't you ask the JIT compiler what it's emitting?

**Comment by borgdylan** [comment]
```uname``` shows OS i.e. Linux/Darwin and the architecture (but only distinguishes between 32 and 64 bit ARM and ignores hard float). ```lscpu``` shows just the CPU stuff and quotes teh architecture like ```uname``` does.
```
dylan@ubuntu-server:~$ lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                16
On-line CPU(s) list:   0-15
Thread(s) per core:    2
Core(s) per socket:    8
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 79
Model name:            Intel(R) Core(TM) i7-6900K CPU @ 3.20GHz
Stepping:              1
CPU MHz:               1200.000
CPU max MHz:           3200.0000
CPU min MHz:           1200.0000
BogoMIPS:              8000.40
Virtualisation:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              20480K
NUMA node0 CPU(s):     0-15
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb intel_pt tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdseed adx smap xsaveopt cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts
```

It is therefore not a viable candidate.

**Comment by borgdylan** [comment]
The internet gave me an answer. If you are on ARM the litmus test is to list all directories under ```/lib``` and see if you have ```arm-linux-gnueabihf``` or ```/lib/arm-linux-gnueabi``` under it.

**Comment by Tragetaschen** [comment]
I will check my platform tomorrow (a "standard" Yocto build) when I have access to the hardware

**Comment by jkotas** [comment]
@marek-safar Do you opinions on the best way to make https://github.com/dotnet/corefx/tree/master/src/System.Runtime.InteropServices.RuntimeInformation work on Mono?

**Comment by Tragetaschen** [comment]
@borgdylan I have checked my embedded system and there isn't any of the two mentioned folders. Might be Debian specific.

```
# find /lib -maxdepth 1 -type d
/lib
/lib/systemd
/lib/modprobe.d
/lib/depmod.d
/lib/udev
/lib/modules
```

However, I do have a `/lib/ld-linux-armhf.so.3` file.

**Comment by borgdylan** [comment]
On x86 I have ```ld-linux.so.2```. I will check later on my pc which is x64 based but has x86 compat libraries as well. Looks like ARM will require different tricks to spot the architecture precisely.

**Comment by borgdylan** [comment]
On systems that have ```gcc``` installed one can use the the ```Target:``` line when invoking:
```
dylan@asusx550ca:~$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/i686-linux-gnu/5/lto-wrapper
Target: i686-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 5.3.1-14ubuntu2.1' --with-bugurl=file:///usr/share/doc/gcc-5/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-5 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-5-i386/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-5-i386 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-5-i386 --with-arch-directory=i386 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-targets=all --enable-multiarch --disable-werror --with-arch-32=i686 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=i686-linux-gnu --host=i686-linux-gnu --target=i686-linux-gnu
Thread model: posix
gcc version 5.3.1 20160413 (Ubuntu 5.3.1-14ubuntu2.1) 
```

**Comment by borgdylan** [comment]
A similar line is output when running ```clang -v```. Using the clang variant may work better on macs since clang is an apple product and is used by apple;s SDKs.

**Comment by Tragetaschen** [comment]
There is only a single C function required from System.Native:

https://github.com/Tragetaschen/libSystem.Native/blob/master/src/GetUnixName.c

And I have wrapped that in an autotools project.

**Comment by borgdylan** [comment]
The API returns more than that. Also it supports getting the architecture. Your solution makes things work but does not yield the expected result.

**Comment by borgdylan** [comment]
Instead I'd wrap the actual code from corefx.

**Comment by Tragetaschen** [comment]
The "code" from corefx is that same C (++) function that instead contains a preprocessor macro whose value comes from the cmake environment. As long as you try to find a solution for Mono an Linux, my code just skips all that boilerplate.

**Comment by borgdylan** [comment]
Corefx calls uname to get a full version string. Any implementation that is to be used in general would need to allow all of the S.R.IS.RI API to work.

**Comment by borgdylan** [comment]
@jkotas can we use the code from corefx to make a reduced version of the library while leaving copyrights and licenses as those of corefx?

**Comment by borgdylan** [comment]
I';d use the builtin gcc macros: https://sourceforge.net/p/predef/wiki/Architectures/

**Comment by jkotas** [comment]
@borgdylan The license describes how you can use this code.

Why you cannot use the System.Native library as is? I understand that it has more stuff in it that what you need to solve your immediate problem, but the extra footprint of this extra stuff should be negligible.

**Comment by Tragetaschen** [comment]
@jkotas 
For me, the entry barrier was really, really high: the immediate problem was the requirement for a llvm/cmake toolchain that targets (my) ARM device. 

cmake is required to get the necessary configuration header files (`pal_config.h`). llvm is required since gcc throws a `sorry, unimplemented: non-trivial designated initializers not supported` for constructs [like this](https://github.com/dotnet/corefx/blob/release/1.0.1/src/Native/System.Native/pal_networking.cpp#L371).

**Comment by borgdylan** [comment]
X86 is not supported correctly by the corefx build system. I had to hack it to get a proper 32 bit library for my nuc and laptop. We need the source to compile cleanly for any architecture that mono builds for.

**Comment by jkotas** [comment]
> We need the source to compile cleanly for any architecture that mono builds for.

We would be happy to take PRs in corefx to make this possible.

> llvm is required since gcc throws a  sorry, unimplemented: non-trivial designated initializers not supported

Again, we would happy to take PRs to make the code compilable with gcc.

> cmake is required to get the necessary configuration header files

This one is harder. It would require introducing parallel redundant build system that is always a problem. We can give it try to see how bad is it if it is something that is really needed to reduce barrier to entry.

**Comment by borgdylan** [comment]
In reality a parallel system would be best as one can build a subset of System.Native directly. All w e need is ```pal_runtimeinformation.cpp``` and ```pal_runtimeinformation.h``` and a new ```lib_runtimeinfo.cpp``` to adapt the gcc/clang macros to do the defines made by cmake. This new makefile would make a different product, a S.R.IS.RI runtime not a .NET Core runtime.

**Comment by borgdylan** [comment]
Using just the files I mentioned would mean that we can get a proper clean x86 build. The current incompatibilities arise form the ```pal_io``` bits which assume that ```int64_t``` and ```__off_t``` are equal. That assumption is true on x64 but not on x86.

**Comment by jkotas** [comment]
I do not think we would want to have parallel build system that builds subsets of the library in the corefx repo.

x86 port is in our roadmap https://github.com/dotnet/core/blob/master/roadmap.md. We just have not gotten to it yet.

**Comment by borgdylan** [comment]
People would like to use Asp.Net Core even where .NET Core cannot run. I would not want to be that person held off from using it to its full potential just because of this.

**Comment by borgdylan** [comment]
I would seek a solution that can be made to work ASAP rather than one that requires waiting for MSFT to finish a certain number of cycles first.

**Comment by borgdylan** [comment]
@jkotas My argument for a separate build script:
.NET Core does aim to support all the architectures that mono supports such as PowerPC and MIPS. There are people that depend on those chips who use mono. Since .NET Core will not run on those architectures, the corefx build system will never support those architectures either.

**Comment by borgdylan** [comment]
Surely no one at MSFT would want to handle all these architectures for all of .NET Core: http://www.mono-project.com/docs/advanced/runtime/#supported-architectures

**Comment by jkotas** [comment]
I have said that we can give the alternative build system try. The System.Native library built via the alternative system should be equivalent to the one built with cmake - it should export same set of methods.

**Comment by borgdylan** [comment]
Could it have a different configuration header? Using the same methods might mean that other OOB assemblies could run on .NET Core and also Mono thru this alternatively built library.

**Comment by borgdylan** [comment]
I will have to use a lot of ```#if defined(...)``` etc. My initial attempt will be to support Linux using g++. When I am done I will invite @Tragetaschen to try building on his ARM board. I will also submit a PR so MSFT could use the CI to make sure that the larger cmake build system does not regress.

**Comment by jkotas** [comment]
> I will have to use a lot of  `#if defined(...)`  etc. 

The preferred way to do ports is to fix the code such that it does not need ifdefs. We can discuss the right approach case by case.

**Comment by borgdylan** [comment]
I will need ifdefs to translate gcc architecture defines to the ones done by cmake.

**Comment by borgdylan** [comment]
First roadblock...g++ will not work. Switching to using clang++.

**Comment by borgdylan** [comment]
Look at: https://github.com/borgdylan/corefx/commit/dbd679a49cdd0c49979682fe29e50b000c6c4c48 This has allowed me to do a build of the whole System.Native without cmake or autotools.

**Comment by borgdylan** [comment]
I will not do any PRs until I have tested S.R.IS.RI against this build on both x64 and x86. The Makefile inside src/Native/System.Native

**Comment by jkotas** [comment]
> without cmake or autotools

Are even autotools barrier to entry? The hardcoded config is pretty non-portable.

**Comment by borgdylan** [comment]
They are not, but they introduce a lot of cruft. For now I attempted to support standard Linux across architectures. More can be added if the contract is expanded.

**Comment by Tragetaschen** [comment]
My 2 cents:

cmake is (at least for me) the much easier requirement and I won't consider it a barrier in a stand-alone project. It's well supported among distributions and meta distribution build tools (like OpenEmbedded) as is in use for several high-profile projects. For my quick fix, I chose autotools just because I used those several years ago and knew where to look.

LLVM on the other hand is not that broadly used on Linux, even less so in the embedded world. So if the standalone library would support GCC, adaption should be considerably easier.

**Comment by jkotas** [comment]
> cmake is (at least for me) the much easier requirement
> LLVM on the other hand is not that broadly used on Linux

This is what I would expect as well. @borgdylan Is there a reason why cmake is a problem for you?

If cmake is fine, fixing the build for gcc and x86 should be pretty clean and non-controversial changes.

**Comment by borgdylan** [comment]
I am not a cmake expert. I would need help to be base stuff on it. As for using GCC, I could try that since I prefer it due to it being supported across all Linux environments my

**Comment by borgdylan** [comment]
The code now builds using g++ instead of clang++ https://github.com/borgdylan/corefx/commit/39e2b71961823f5d96bf3349b7798a93dea8a17e

**Comment by jkotas** [comment]
The changes to make it build with g++ are good. Would you like to cherry pick them into a PR? (Multiple smaller PRs are easier to deal with - get through.)

**Comment by borgdylan** [comment]
I'll try once I arrive at work. It's 7:30 am here.

**Comment by borgdylan** [comment]
So the PR got accepted. I will make a new branch with the standalone build based on the PR and @jkotas 's feedback on the current version.

**Comment by borgdylan** [comment]
@jkotas @Tragetaschen @ericstj The new standalone stuff will be done here: https://github.com/borgdylan/corefx/commits/sysnative_standalone2

**Comment by borgdylan** [comment]
With my latest round of testing I can conclude that the library built by the standalone and normal build systems lead to the same behaviour of the S.R.IS.RI API under mono.

**Comment by migueldeicaza** [comment]
Hello

We just released a beta of Mono with support for .NET Standard 1.6, would you guys be so kind to test again?

I tried using dotnet core, but it does not seem to be setup to work with those of us that do not want to have a system-gloabl brew installation.

**Comment by zgramana** [comment]
For anyone looking for a workaround (on macOS/Darwin at least), here are my notes on getting things working on OS X (and in theory they should work on Ubuntu with little or no modification, but I haven't tested):

1. **Find your loadable native images** (`.dylib` on macOS/Darwin, `.so` on Linux). The location depends on what your dev/packaging platform is.
 1. If you're building and running on the same platform, you can likely find them at `/usr/local/share/dotnet/shared/Microsoft.NETCore.App/1.0.0`.
 1. If you're cross-compiling, or want to be more generalized, you can find them for each installed platform using this command:
  ```
   ls ~/.nuget/packages/runtime.*.*.*.runtime.native.System/1.0.1*/runtimes/*/native
  ```
1. **Give your native images loadable filenames**. The loader function, `dlopen`, is going to probe the filesystem for a file prefixed with `lib`. Even if it knows *where* to find your images, without the prefix, `dlopen` will not find them. If you want to watch this process happen, just prefix the `dotnet` command like this:
  ```
   MONO_LOG_LEVEL=debug dotnet run
  ```
  You'll see all the probing going one, like:
  ```
   Mono: DllImport error loading library '/path/to/your/src/project/bin/Debug/net451/osx.10.11-x64/libSystem.Native.dylib': 'dlopen(/path/to/your/src/project/bin/Debug/net451/osx.10.11-x64/libSystem.Native.dylib, 9): image not found'.
  ```
 1. **Rename/Copy/Symlink from `System.*.{dylib|so}` to `libSystem.*.{dylib|so}`**.
 1. **Map the expected name to the actual name**.  This adds an rpath that lets `dlopen` see `lib{FOO}` even though the filesystem only has {FOO}:
 ```
  sudo install_name_tool -add_rpath /usr/local/share/dotnet/shared/Microsoft.NETCore.App/1.0.0/libSystem.Native.dylib /usr/local/share/dotnet/shared/Microsoft.NETCore.App/1.0.0/System.Native.dylib
 ```
 **Notes**: On Linux, looks like you'd probably use [patchelf](http://nixos.org/patchelf.html) instead. For either option, you can automate this process for all your files with a little `awk` fu:
  ```
   sudo ls /usr/local/share/dotnet/shared/Microsoft.NETCore.App/1.0.0/System*.dylib | awk '{split($1,arr,"/");name=arr[length(arr)];i=index($1, name);path=substr($1, 0, i - 1);newName=path "lib" name;cmd="sudo install_name_tool -add_rpath " newName " "  $1;cmd | getline; print}'
  ```
1. [Only if you chose 2.i] **Put your native images in `dlopen`'s search path**.  You have a few choices, and depends on what your ultimate goals are.
 1. Copy the respective `libSystem.Native.{dylib|so}` files from *Step 2* to `bin/{config}/{framework}/{platform}/libSystem.Native.{dylib|so}`. Nothing fancy/brute force. Do this for each project. If someone does a clean clone of a repo, or you're doing stuff in CI, you should really add a pre-compile task to get this done. Be mindful of any tooling/automation that might periodically `rm -Rf bin/Debug/*`.
 1. Set `LD_LIBRARY_PATH` to one or more paths containing your native images. Normally on macOS/Darwin this would be `DYLD_LIBRARY_PATH` and it's ilk (see [`man DYLD_LIBRARY_PATH`](https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man1/dyld.1.html) for the details), but seems Mono fine with either one. I don't know if they ever plan to change that, so YMMV.

#### Notes
Regardless of your choice in *Step 3*, only `libuv.{dylib|so}` will make into the published folder. 
For files like `libSystem.Native` you'll have to copy that into the `publish` folder yourself...unless you already installed the .NET Core on the machine you're deploying too. In that case, you could wrap your executable in a simple one-line shell script that looked something like:  `LD_LIBRARY_PATH=/usr/local/share/dotnet/shared/Microsoft.NETCore.App/1.0.0 mono bin/Debug/net451/osx.10.11-x64/my-app.exe`. If you were to symlink `Microsoft.NETCore.App/Current` to `Microsoft.NETCore.App/1.0.0` you could reasonably easily update without touching your script.

**Comment by sunsided** [comment]
Btw., for the case of the missing System.Native library, https://github.com/aspnet/KestrelHttpServer/issues/963#issuecomment-241299565 provides a workaround. I managed to get `net4xx` targets working with that on mono 4.7 using the `3131` tooling.

**Comment by AlexGhiondea** [comment]
@jkotas @ericstj @Priya91 what are the next steps for this?

**Comment by DustinCampbell** [comment]
What's the plan to address this? We just discovered that this package slipped into the desktop C# compiler causing it to no longer run on Mono.

cc @jaredpar

**Comment by jkotas** [comment]
@DustinCampbell Desktop != Mono. The desktop C# compiler package is meant to run on desktop. If you run app build for desktop on Mono, your millage will vary. That's by design.

**Comment by agocke** [comment]
@jkotas I thought the point of "netstandard" was that it would allow people to build different runtimes targeting the same standard.

If Mono is an implementation of "net46" how can it not be expected to be compatible with netstandard libraries compiled against "net46"?

**Comment by davidfowl** [comment]
That's true but this package *should* be made to work on mono. @emgarten do we have a mono TFM as yet?

**Comment by jkotas** [comment]
> mono TFM

Yes, mono TFM is the proper "correct by construction" solution for this problem.

**Comment by DustinCampbell** [comment]
agreed -- for many things. The Microsoft.Build.Runtime package would benefit greatly from that TFM as well

**Comment by agocke** [comment]
Maybe for some things. But that abandons the notion that Mono is an alternate implementation of some/most of net46 -- which its entire purpose today.

If you abandon that design, you are leaving behind almost all of Mono's consumers.

**Comment by emgarten** [comment]
> do we have a mono TFM as yet? 

there is no mono TFM in NuGet currently

**Comment by akoeplinger** [comment]
@jkotas 

> It has not caught up with System.Runtime.InteropServices.RuntimeInformation yet. 

That's not true, we've been shipping an implementation for that in Mono 4.6. The issue is that the System.Runtime.InteropServices.RuntimeInformation nuget package contains an implementation in the lib\net45 folder which will be copied to the app directory and since it calls into Windows APIs it'll crash on other platforms.

Mono prefers loading an assembly from the app directory, that's why the GAC version is not getting used. The PR you mentioned above works around that problem by blacklisting assemblies which we know to be broken on Mono, which is hardly a perfect solution. We're discussing the larger issue at https://github.com/dotnet/corefx/issues/15112 and should probably continue the discussion there.

**Comment by jkotas** [comment]
Agree. I should have been more specific - Mono has not caught up yet with how System.Runtime.InteropServices.RuntimeInformation and similar OOB packages are deployed in desktop apps.

**Comment by davidfowl** [comment]
@akoeplinger what version of mono is going to fix this?

**Comment by akoeplinger** [comment]
@davidfowl we shipped a "fix" with Mono 4.8 which essentially blacklists some specific assemblies+versions like SRI.RuntimeInformation: https://github.com/mono/mono/blob/mono-4.8.0-branch/mono/metadata/image.c#L1035-L1091

A more future-proof solution is currently being discussed in #15112

**Comment by Daniel15** [comment]
For what it's worth, even with Mono 5.4 I still have to symlink `/usr/local/bin/libSystem.Native.so` to `System.Native.so` in order to get ASP.NET Core 2.0 sites to load.
